upup/pkg/fi/cloudup/template_functions.go

```diff
@@ -272,6 +275,29 @@ func (tf *TemplateFunctions) GetInstanceGroup(name string) (*kops.InstanceGroup,
 	return ig, nil
 }
 
+// ControlPlaneControllerReplicas returns the amount of replicas for a controllers that should run in the cluster
+// If the cluster has a highly available control plane, this function will return 2, if it has 1 control plane node, it will return 1
+func (tf *TemplateFunctions) ControlPlaneControllerReplicas() int {
+	if tf.HasHighlyAvailableControlPlane() {
+		return 2
```

Comment:
Perhaps this should take the `2` as an input parameter? Return the max of input parameter and number of API servers?

Reply:
I'm trying to figure out the benefit of having this sort of logic inside the template functions, versus having a template function for returning the number of apiservers and leaving the logic in go template.

Reply:
I think in almost all cases one wants this to be 2 for HA as these are leader elected.

This is a somewhat simple function, but I think it would be used a lot, so it will reduce some boilerplate.

Reply:
Perhaps. A lot of these things are one-per-apiserver, but that's probably for other reasons. We might want to give weak zone anti-affinity for the 5-apiservers-in-3-zones case.

Reply:
I don't think there are any good arguments for that. We have the same discussions around KCM, CCM etc .. these leader-elected controllers only run a single active instance and having two standbys is not very useful.

There are other components that are webhooks (or like metrics-server) and could be scaled using the autoscaler we use for the kube-dns service, but they often do not have to run on CP nodes at all.

But the leader-elected scenario is why I named the functions like I did as well.

Reply:
Also worth mentioning that the PDB in this case guarantees that there will be an active leader at any point in time (given the PDB is respected)

Reply:
Yes, I am convinced two are sufficient for leader-elected controllers, provided such controllers have some form of zone anti-affinity (when there are multiple zones).

As to naming, there's the ambiguity of what a "control plane node" is when etcd and apiserver are split. But that's a different issue—it would not make sense to have HA etcd with non-HA apiserver or vice versa.

Reply:
Good point on topology key. For CP nodes that's the same, but with IRSA we have to remember to change this to zone.

"control plane node" are those with the role "control-plane", which we in an office hour informally concluded to be "what is running KCM, kube-scheduler, and similar". Doesn't necessarily have to run etcd. Unknown what this means in the long-term. Different discussion though.